c++/src/capnp/compiler/capnpc-c++.c++

```diff
@@ -2188,7 +2188,12 @@ private:
   InterfaceText makeInterfaceText(kj::StringPtr scope, kj::StringPtr name, InterfaceSchema schema,
                                   kj::Array<kj::StringTree> nestedTypeDecls,
                                   const TemplateContext& templateContext) {
-    auto fullName = kj::str(scope, name, templateContext.args());
+    auto templateArgs = templateContext.args();
+    auto fullName = kj::str(scope, name, templateArgs);
+    auto fullDisambiguatedName = kj::str(
+        scope.size() ? "typename " : "", scope,
+        scope.size() && templateArgs.size() ? "template " : "", name,
```

Comment:
The lack of more `template` sprinkles might be an oversight: it looks like all the templates this code affects in test.capnp.h are all only nested one-deep.

I'll try using the unqualified parent type name first and remove this stuff. If that doesn't work for some reason, I'll add a test case with multiple nested generics to elicit the problem you described, then fix it.